src/Components/DatePicker/Component.purs

```diff
@@ -0,0 +1,428 @@
+import Data.Either (either)
+import Data.Formatter.DateTime (formatDateTime)
+import Data.Maybe (Maybe(..), fromMaybe, maybe)
+import Data.String (trim)
+import Data.Tuple (Tuple(..), fst, snd)
+import Effect.Aff.Class (class MonadAff)
+import Effect.Class.Console (log)
+import Effect.Now (nowDate)
+import Halogen as H
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties as HP
+import Ocelot.Block.Button as Button
+import Ocelot.Block.Format as Format
+import Ocelot.Block.Icon as Icon
+import Ocelot.Block.Input as Input
+import Ocelot.Components.DatePicker.Utils as Utils
+import Ocelot.Data.DateTime as ODT
+import Ocelot.HTML.Properties (css)
+import Select as Select
+import Select.Utils.Setters as Setters
+import Web.Event.Event (preventDefault)
+import Web.UIEvent.KeyboardEvent (KeyboardEvent)
+import Web.UIEvent.KeyboardEvent as KE
+
+type State =
+  { targetDate :: Tuple Year Month
+  , selection :: Maybe Date
+  , search :: String
+  }
+
+type Input =
+  { targetDate :: Maybe (Tuple Year Month)
+  , selection :: Maybe Date
+  }
+
+data Query a
+  = HandleSelect (Select.Message Query CalendarItem) a
+  | ToggleYear  Direction a
+  | ToggleMonth Direction a
+  | Initialize a
+  | TriggerFocus a
+  | Synchronize a
+  | GetSelection (Maybe Date -> a)
+  | SetSelection (Maybe Date) a
+  | Key KeyboardEvent a
+  | Search String a
+  | Receive Input a
+
+data Message
+  = SelectionChanged (Maybe Date)
+  | VisibilityChanged Select.Visibility
+  | Searched String
+
+data Direction
+  = Prev
+  | Next
+
+type ParentHTML m
+  = H.ParentHTML Query ChildQuery Input m
+
+type ChildSlot = Unit
+type ChildQuery = Select.Query Query CalendarItem
+
+----------
+-- Calendar Items
+
+data CalendarItem
+  = CalendarItem SelectableStatus SelectedStatus BoundaryStatus Date
+
+data SelectableStatus
+  = NotSelectable
+  | Selectable
+
+data SelectedStatus
+  = NotSelected
+  | Selected
+
+data BoundaryStatus
+  = OutOfBounds
+  | InBounds
+
+dropdownClasses :: Array HH.ClassName
+dropdownClasses = HH.ClassName <$>
+  [ "absolute"
+  , "shadow"
+  , "pin-t"
+  , "pin-l"
+  , "z-50"
+  , "border"
+  , "border-grey-90"
+  , "rounded"
+  , "p-6"
+  , "bg-white"
+  , "text-center"
+  , "text-lg"
+  ]
+
+component :: ∀ m
+  . MonadAff m
+ => H.Component HH.HTML Query Input Message m
+component =
+  H.lifecycleParentComponent
+    { initialState
+    , render
+    , eval
+    , receiver: HE.input Receive
+    , initializer: Just $ H.action Initialize
+    , finalizer: Nothing
+    }
+  where
+    initialState :: Input -> State
+    initialState { targetDate, selection } =
+      { targetDate: fromMaybe (Tuple (ODT.unsafeMkYear 2001) (ODT.unsafeMkMonth 1)) targetDate
+      , selection
+      , search: ""
+      }
+
+    eval
+      :: Query
+      ~> H.ParentDSL State Query ChildQuery Unit Message m
+    eval = case _ of
+      Search text a -> do
+        H.raise $ Searched text
```

Comment:
Raising a message causes the evaluation of that raise to occur first. So if raising this message results in the parent performing some expensive computation it will delay everything from here onward until that completes.

In most cases raising the message should be the last effect in the `eval` because of this. Unless there's a particular reason to have it first, I'd recommend moving this to the end of the code block.